tools/net-common.in

```diff
@@ -63,6 +63,41 @@ function fix_mac {
   ip link set $INTERFACE address $FIXED_MAC || true
 }
 
+function bridge_vlan_add {
+  local VID="${1}"
+  shift
+  local EXTRA="${@}"
+  # add the VLAN to the tap
+  bridge vlan add dev ${INTERFACE} vid ${VID} ${EXTRA}
+  # add the VLAN to the bridge uplink
+  bridge vlan add dev ${BRIDGE_UPLINK} vid ${VID}
+}
+
+function setup_bridge_vlan_aware {
+  BRIDGE_UPLINK="$(bridge link show | grep "master ${LINK} state" | grep -v tap | cut -d' ' -f2)"
```

Comment:
`BRIDGE_UPLINK` silently assumes that there will be only one non-tap interface connected to the bridge, which might not always be the case. If there are multiple non-tap interfaces, `BRIDGE_UPLINK` will end up with space-delimited values, breaking `bridge_vlan_add` subsequently. Now, since there is no hard concept of "uplink port" in the Linux bridge and since we don't place any special constraints wrt. network topology for running ganeti, we should probably add the VLAN on *all* non-tap interfaces.

By the way, I would be very happy if we could somehow discern instance interfaces without relying on the interface name (using `ip link set ... group ganeti` could work), but this is outside the scope of the current PR.

Furthermore, as far as I can tell, we don't check that the bridge has VLAN filtering enabled. We should at least check sysfs to see if the bridge has vlan_filtering enabled, and fail otherwise. This will be a good line of defense against misconfigurations (e.g. VLAN set, but filtering not enabled, or trying to add a VLAN-enabled instance to a "traditional" ganeti setup which only has access ports on the bridge).